lib/pure/ospaths.nim

```diff
@@ -157,64 +158,173 @@ const
     ## The character which separates the base filename from the extension;
     ## for example, the '.' in ``os.nim``.
 
+# TODO: MOVE to strutils, sequtils, or algorithms
+proc countWhile*(str: string, pred : proc(a:char):bool, prefix = true): int =
+  ## returns the number of elements in ``str`` that satisfy ``pred`` predicate.
+  ## without interruption, starting from beginning (when ``prefix`` is true)
+  ## or end (when ``prefix`` is false)
+  runnableExamples:
+    import sugar
+    doAssert countWhile("abc", a=>a == '/') == 0
+    doAssert countWhile("//abc", a=>a == '/') == 2
+    doAssert countWhile("//", a=>a == '/') == 2
+    doAssert countWhile("", a=>a == '/') == 0
+
+    doAssert countWhile("abc", a=>a == '/', prefix = false) == 0
+    doAssert countWhile("abc//", a=>a == '/', prefix = false) == 2
+    doAssert countWhile("//", a=>a == '/', prefix = false) == 2
+    doAssert countWhile("", a=>a == '/', prefix = false) == 0
+  let lenS = len(str)
+  if prefix:
+    for i in 0..<lenS:
+      if not pred(str[i]): return i
+  else:
+    for i in 0..<lenS:
+      if not pred(str[lenS - i - 1]): return i
+  return lenS
+
+# TODO: once slices are supported, define instead `rootPrefix`
+proc rootPrefixLength*(path: string) : int =
```

Comment:
Once slices are supported? Just return the prefix and call ``len`` on it.

Reply:
`rootPrefixLength` avoids allocating. If/when we have slices, this wouldn't be a problem; until then it is.